src/Http/Routing/src/Matching/HttpMethodDestinationsLookup.cs

```diff
@@ -0,0 +1,131 @@
+                        {
+                            putDestination = destination;
+                            continue;
+                        }
+                        else if (method.Equals(HttpMethods.Post, StringComparison.OrdinalIgnoreCase))
+                        {
+                            postDestination = destination;
+                            continue;
+                        }
+                        else if (method.Equals(HttpMethods.Patch, StringComparison.OrdinalIgnoreCase))
+                        {
+                            patchDestination = destination;
+                            continue;
+                        }
+                        break;
+                    case 't' when method.Equals(HttpMethods.Trace, StringComparison.OrdinalIgnoreCase):
+                        traceDestination = destination;
+                        continue;
+                }
+            }
+
+            _extraDestinations ??= new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+            _extraDestinations.Add(method, destination);
+        }
+
+        _connectDestination = connectDestination ?? _exitDestination;
+        _deleteDestination = deleteDestination ?? _exitDestination;
+        _getDestination = getDestination ?? _exitDestination;
+        _headDestination = headDestination ?? _exitDestination;
+        _optionsDestination = optionsDestination ?? _exitDestination;
+        _patchDestination = patchDestination ?? _exitDestination;
+        _postDestination = postDestination ?? _exitDestination;
+        _putDestination = putDestination ?? _exitDestination;
+        _traceDestination = traceDestination ?? _exitDestination;
+    }
+
+    public int GetDestination(string method)
+    {
+        // Implementation skeleton taken from https://github.com/dotnet/runtime/blob/13b43155c31beb844b1b04766fea65235ccd8363/src/libraries/System.Net.Http/src/System/Net/Http/HttpMethod.cs#L179
+        if (method.Length >= 3) // 3 == smallest known method
+        {
+            (var matchedMethod, var destination) = (method[0] | 0x20) switch
+            {
+                'c' => (HttpMethods.Connect, _connectDestination),
+                'd' => (HttpMethods.Delete, _deleteDestination),
+                'g' => (HttpMethods.Get, _getDestination),
+                'h' => (HttpMethods.Head, _headDestination),
+                'o' => (HttpMethods.Options, _optionsDestination),
+                'p' => method.Length switch
+                {
+                    3 => (HttpMethods.Put, _putDestination),
+                    4 => (HttpMethods.Post, _postDestination),
+                    _ => (HttpMethods.Patch, _patchDestination),
+                },
+                't' => (HttpMethods.Trace, _traceDestination),
+                _ => (null, 0),
+            };
+
+            if (matchedMethod is not null && method.Equals(matchedMethod, StringComparison.OrdinalIgnoreCase))
```

Comment:
I'm wondering if we can speed-up the `method.Equals(matchedMethod, StringComparison.OrdinalIgnoreCase)` since we know we are targeting a well-known method. Maybe we can leverage the fact that it's very likely ASCII on the header to perform two memory comparisons and or the results for the individual characters?

Reply:
Do you mean to compare with upper-case and lower-case, OR the results and check if all values set to 1? I actually wonder if there is something like this in System.Text.Ascii already that I could use straight away. And I assume we cannot expect it as ASCII, so if the match fails, fall back to regular string comparison, right?

Reply:
Yeah, there is potential to speed this up:
* One option is to generate a trie to compare the strings. That's probably overkill.
* A simpler optimization would be to have a fast path that assumes upper case method (which basically everyone uses) and compares against hardcoded chars using `SequenceEqual`. If it doesn't match then fallback to string.Equals + OrdinalIgnoreCase.

However, this PR has a nice improvement and sets up good infrastructure. Merge now and iterate in the future.

Reply:
I would have the two versions of the method in a buffer, (upper and lowercase), compare against it and or the results. If done with simd it's about 4 ops.

Disclaimer that I haven't done it myself, but I know in other places we use things like SWAR (Simd within a register) so it's likely useful to apply it here too, rely on SIMD instructions directly or use any specialized helper.

Reply:
I am happy to come explore @javiercn , I have written some vectorized code in the past, let me see if I can do something "simple". But Ascii type also uses [vectorization](https://source.dot.net/#q=Ascii.EqualsIgnoreCase).

Reply:
> I would have the two versions of the method in a buffer, (upper and lowercase), compare against it and or the results. If done with simd it's about 4 ops.

It would be a bit more complicated because mix case is valid, e.g. `Post`.

Reply:
@JamesNK  and @javiercn I promised to come back to this thread. TLDR; none of the above suggestions were performing better to what is already merged.

## Using ASCII

Relevant code:

```csharp
if (matchedMethod is not null && (System.Text.Ascii.EqualsIgnoreCase(matchedMethod, method) || method.Equals(matchedMethod, StringComparison.OrdinalIgnoreCase)))
{
    return destination;
}
```

ASCII Results

```
|                     Method | TestHttpMethod |      Mean |     Error |    StdDev |    Median |          Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |--------------- |----------:|----------:|----------:|----------:|--------------:|------:|------:|------:|----------:|
|  DictionaryPolicyJumpTable |            GET |  6.182 ns | 0.0793 ns | 0.0703 ns |  6.147 ns | 161,768,263.8 |     - |     - |     - |         - |
| SingleEntryPolicyJumpTable |            GET |  2.721 ns | 0.0557 ns | 0.1344 ns |  2.674 ns | 367,483,825.2 |     - |     - |     - |         - |
|  DictionaryPolicyJumpTable |          Merge | 12.403 ns | 0.1286 ns | 0.1140 ns | 12.396 ns |  80,624,929.0 |     - |     - |     - |         - |
| SingleEntryPolicyJumpTable |          Merge |  3.363 ns | 0.0221 ns | 0.0196 ns |  3.363 ns | 297,385,268.2 |     - |     - |     - |         - |
|  DictionaryPolicyJumpTable |           POST |  6.330 ns | 0.0457 ns | 0.0405 ns |  6.336 ns | 157,968,543.3 |     - |     - |     - |         - |
| SingleEntryPolicyJumpTable |           POST |  3.442 ns | 0.0702 ns | 0.0721 ns |  3.444 ns | 290,548,813.4 |     - |     - |     - |         - |
```

## Using SequenceEquals Ordinal comparison

Relevant code:

```csharp
if (matchedMethod is not null && (method.AsSpan().SequenceEqual(matchedMethod)
        || method.Equals(matchedMethod, StringComparison.OrdinalIgnoreCase)))
{
    return destination;
}
```

SequenceEqual Results

```
|                     Method | TestHttpMethod |      Mean |     Error |    StdDev |          Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |--------------- |----------:|----------:|----------:|--------------:|------:|------:|------:|----------:|
|  DictionaryPolicyJumpTable |            GET |  5.325 ns | 0.0951 ns | 0.0794 ns | 187,810,373.8 |     - |     - |     - |         - |
| SingleEntryPolicyJumpTable |            GET |  3.136 ns | 0.0436 ns | 0.0387 ns | 318,917,169.1 |     - |     - |     - |         - |
|  DictionaryPolicyJumpTable |          Merge | 12.382 ns | 0.1478 ns | 0.1383 ns |  80,761,722.6 |     - |     - |     - |         - |
| SingleEntryPolicyJumpTable |          Merge |  3.360 ns | 0.0353 ns | 0.0330 ns | 297,646,602.7 |     - |     - |     - |         - |
|  DictionaryPolicyJumpTable |           POST |  4.894 ns | 0.0540 ns | 0.0505 ns | 204,328,749.4 |     - |     - |     - |         - |
| SingleEntryPolicyJumpTable |           POST |  3.343 ns | 0.0078 ns | 0.0065 ns | 299,173,275.5 |     - |     - |     - |         - |
```

## Vector

A slightly different implementation I take more optimistic approach that only compares with upper-case vectorized, and if no match falls back to regular approach. The problem with vectors is that they are a bit more difficult to create when the size of the vector is larger than the input. 

So what I ended up doing is creating vectorized versions of the known HTTP verbs and a vector mask corresponding to their length.

```csharp
       // Set values to vConnect, vDelete, vGet etc. fields in ctor

        vConnect = Vector256.LoadUnsafe<ushort>(ref Unsafe.As<char, ushort>(ref MemoryMarshal.GetReference(HttpMethods.Connect.AsSpan()))) & vMask7;
        vDelete = Vector256.LoadUnsafe<ushort>(ref Unsafe.As<char, ushort>(ref MemoryMarshal.GetReference(HttpMethods.Delete.AsSpan()))) & vMask6;
        vGet = Vector256.LoadUnsafe<ushort>(ref Unsafe.As<char, ushort>(ref MemoryMarshal.GetReference(HttpMethods.Get.AsSpan()))) & vMask3;
        // ...
        vMask3 = new Vector<ushort>(new ushort[16] { ushort.MaxValue, ushort.MaxValue, ushort.MaxValue, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 }).AsVector256();
        vMask4 = new Vector<ushort>(new ushort[16] { ushort.MaxValue, ushort.MaxValue, ushort.MaxValue, ushort.MaxValue, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 }).AsVector256();
         //...
    }

    public int GetDestination(string method)
    {
        if (Vector256.IsHardwareAccelerated)
        {
            (var vMatched, var vMask, var expectedLength, var destination) = (method[0]) switch
            {
                'C' => (vConnect, vMask7, 7, _connectDestination),
                'D' => (vDelete, vMask6, 6, _deleteDestination),
                'G' => (vGet, vMask3, 3, _getDestination),
                'H' => (vHead, vMask4, 4, _headDestination),
                'O' => (vOptions, vMask7, 7, _optionsDestination),
                'P' => method.Length switch
                {
                    3 => (vPut, vMask3, 3, _putDestination),
                    4 => (vPost, vMask4, 4, _postDestination),
                    _ => (vPatch, vMask5, 5, _patchDestination),
                },
                'T' => (vTrace, vMask5, 5, _traceDestination),
                _ => (Vector256<ushort>.Zero, Vector256<ushort>.Zero, 0, 0),
            };

            if (expectedLength == method.Length)
            {
                ref var source = ref Unsafe.As<char, ushort>(ref MemoryMarshal.GetReference(method.AsSpan()));
                var vSource = Vector256.LoadUnsafe<ushort>(ref source);
                vSource = vSource & vMask;

                if (Vector256.Equals(vSource, vMatched) == Vector256<ushort>.AllBitsSet)
                {
                    return destination;
                }
            }
        }

       // else fallback to non-vectorized approach

}
```

I was considering to implement the lower-case comparison too but given the performance tests indicate this solution is already slower to the `string.Equals` even when the input is upper cased, extending it won't get any faster.

Vectorized Results:

```
|                     Method | TestHttpMethod |      Mean |     Error |    StdDev |    Median |          Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------- |--------------- |----------:|----------:|----------:|----------:|--------------:|------:|------:|------:|----------:|
|  DictionaryPolicyJumpTable |            GET |  4.669 ns | 0.0417 ns | 0.0369 ns |  4.669 ns | 214,196,679.8 |     - |     - |     - |         - |
| SingleEntryPolicyJumpTable |            GET |  3.119 ns | 0.0193 ns | 0.0181 ns |  3.116 ns | 320,565,879.3 |     - |     - |     - |         - |
|  DictionaryPolicyJumpTable |          Merge | 13.416 ns | 0.1905 ns | 0.1688 ns | 13.446 ns |  74,538,092.7 |     - |     - |     - |         - |
| SingleEntryPolicyJumpTable |          Merge |  3.375 ns | 0.0281 ns | 0.0249 ns |  3.367 ns | 296,253,388.9 |     - |     - |     - |         - |
|  DictionaryPolicyJumpTable |           POST |  4.382 ns | 0.0920 ns | 0.1682 ns |  4.292 ns | 228,209,040.8 |     - |     - |     - |         - |
| SingleEntryPolicyJumpTable |           POST |  3.364 ns | 0.0283 ns | 0.0237 ns |  3.374 ns | 297,300,993.8 |     - |     - |     - |         - |
```